# WORKSHOP - ESTATÍSTICA BÁSICA COM R - GRUPO TURING

**Tabela de Conteúdos**

- :open_file_folder: [Importando dados em R](#parte1)
    - [Importando um csv](#importando_csv)
    - [Boas práticas em criação de bases de dados](#boas_praticas)
    - [Importando os dados do Workshop](#importando_ws)
    <br><br>
- :mag: [Estatística básica em R](#parte2)
    - [Tipos de dados](#tipos_de_dados)
        - [Dados Qualitativos](#quali)
        - [Dados Quantitativos](#quant)
    - [Estatística Descritiva](#descritiva)
        - [Medidas de tendência central](#tendencia_central)
            - [Média](#media)
            - [Mediana](#mediana)
            - [Percentis](#percentil)
            - [Summary](#summary)
        - [Medidas de variação uni e multivariadas](#variacao)
            - [Amplitude](#amplitude)
            - [Amplitude interquartil](#amplitude_interquartil)
            - [Variância](#variancia)
            - [Desvio Padrão](#desv_pad)
            - [Covariância](#covar)
            - [Coeficiente de Correlação](#ccor)
            <br><br>
- :bookmark_tabs: [Manipulando dados em R](#parte3)
    - [Dados Faltantes](#dados_faltantes)
    - [Deep dive in dplyr](#dplyr)
        - [mutate() - criando novas variáveis](#mutate)
        - [select() - selecionando colunas](#select)
        - [filter() - filtrando a base de dados](#filter)
        - [summarise() - a tabela dinâmica do R](#summarise)
        - [arrange() - alterando a ordem das linhas](#arrange)
        - [Combinando os verbos](#varios_verbos)
        <br><br>
- :chart_with_downwards_trend: [Visualizando dados em R](#parte4)
<br><br>
- :question: [Inferência Estatística em R](#parte5)


## <a id = "parte1"></a> <span style="color: orange;"> Parte 1 - Importando dados em R </span>

Quando vamos fazer nossa análise de dados, comumente devemos trazer esses dados de algum outro arquivo, o que nos pede um pouco de cuidado. <br> Por exemplo, é importante sabermos qual é o tipo de arquivo que estamos importando, pois isso pode fazer nosso programa ler os dados errôneamente. <br> Um dos formatos mais comums para a análise de dados é o **CSV**, ou *Comma separeted values*.

Assim, se for salvar uma planilha de dados, não se esqueça de salvar como CSV.

![img](img/csv_meme.jpg)

Outro fator importante que devemos levar em conta na hora de importar o arquivo é o local em que ele está salvo. Se ele for salvo no mesmo diretório que o programa que se está usando para ler os dados, ou alguma pasta dentro deste, isso acarretará que podemos apenas escrever o nome do arquivo, sem a necessidade de escrever todo o caminho. <br> Vamos ver um exemplo abaixo:

### <a id="importando_csv"></a>Importando um arquivo CSV 

```R
mydata <- read.csv("Data.csv", header=TRUE)
```

Aqui ```mydata``` é o nome da variável que atribuiremos o conteúdo do arquivo que estamos lendo <br>
A função ```read.csv``` lê arquivos que possuem este formato, aí uma importância de sabermos o formato do arquivo que estamos trabalhando. <br>
O nome do arquivo deve vir dentro de aspas (simples ou dupla). Neste caso o arquivo está no mesmo diretório em que o nosso programa está sendo salvo, por isso não há necessidade de especificar diretórios anteriores. <br>
Um argumento diferente da função deve ser separado por vírgula, aqui colocamos a função ```header=TRUE``` que significa que a primeira linha do nosso arquivo contém o nome das colunas (*labels*). Caso a primeira linha já contenha dados, vc deve trocar esse argumento por ```header=FALSE``` .

Caso o arquivo esteja em um diretório diferente, é necessário identificá-lo. Vamos ver um exemplo:

```R
mydata <- read.csv("C:/Usuarios/Caminho/do/arquivo/Data.csv", header=TRUE)
```

### Alguns problemas 

É comum importarmos um arquivo e este não parecer estar com um formato adequado, como por exemplo uma coluna sem dados, ou duplicada. Nestes casos, é importante checar se os dados estão da mesma forma que seu programa está lendo. Uma forma de descobrir isso é abrir os dados em um programa como o excel, ou leitor de texto e verificar se a divisão das colunas está sendo feita por vírgula ```,``` ou ponto e vírgula ```;```.

Caso o separador não seja a vírgula, podemos usar o argumento  ```sep=" "``` como o separador desejado:

``` R 
mydata4 <- read.csv("Data3.csv", header=TRUE, sep=";")
```

## <a id="boas_praticas"></a> Boas Práticas 

Caso você não queira ter problemas na hora de importar dados para fazer suas análises, são necessárias algumas boas práticas na hora de criar seu arquivo.

Neste WorkShop, usaremos o dataset Adults, que possui dados do censo norte americano sobre fatores socioeconômicos. Vamos visualizá-lo no Excel para entender sua estrutura:

![img2](img/excel_adults.png)

Datasets também podem ser montados num simples arquivo de texto:

```
Id,age,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
16280,34,Divorced,Exec-managerial,Own-child,White,Male,0,0,44,United-States,<=50K
16281,58,Married-civ-spouse,Transport-moving,Husband,White,Male,0,0,40,United-States,<=50K
```

### Como escrever então seu arquivo de dados?

1. Comece pela primeira linha e primeira coluna, ou seja, não pule células
2. Inclua uma linha de títulos para cada coluna
3. Tenha certeza que o número de títulos é igual ao número de colunas dos seus dados
4. Não junte células, isso pode fazer seu programa ter problemas na hora de ler o arquivo
5. No caso de arquivos de texto, use preferencialmente ```,``` como separador dos dados

## <a id="importando_ws"></a> Importando o arquivo do Workshop 

In [1]:
mydata <- read.csv("dados/adults.csv", header=TRUE)
head(mydata, 5)

Id    age workclass    fnlwgt education    education.num marital.status    
1 16280 34  Private      204991 Some-college 10            Divorced          
2 16281 58  Local-gov    310085 10th          6            Married-civ-spouse
3 16282 25  Private      146117 Some-college 10            Never-married     
4 16283 24  Private      138938 Some-college 10            Divorced          
5 16284 57  Self-emp-inc 258883 HS-grad       9            Married-civ-spouse
  occupation        relationship  race  sex    capital.gain capital.loss
1 Exec-managerial   Own-child     White Male      0         0           
2 Transport-moving  Husband       White Male      0         0           
3 Machine-op-inspct Not-in-family White Male      0         0           
4 Adm-clerical      Not-in-family White Female    0         0           
5 Transport-moving  Husband       White Male   5178         0           
  hours.per.week native.country income
1 44             United-States  <=50K 
2 40             United-States  <=50K 
3 42             United-States  <=50K 
4 40             United-States  <=50K 
5 60             Hungary        >50K

Note que por padrão, R lê seu arquivo CSV como se fosse separado por ```,``` <br>
O que acontece então se usarmos outro separador?

In [2]:
my_wrong_data <- read.csv("dados/adults.csv", header=TRUE, sep=";")
head(my_wrong_data, 5)

Id.age.workclass.fnlwgt.education.education.num.marital.status.occupation.relationship.race.sex.capital.gain.capital.loss.hours.per.week.native.country.income
1 16280,34,Private,204991,Some-college,10,Divorced,Exec-managerial,Own-child,White,Male,0,0,44,United-States,<=50K                                              
2 16281,58,Local-gov,310085,10th,6,Married-civ-spouse,Transport-moving,Husband,White,Male,0,0,40,United-States,<=50K                                            
3 16282,25,Private,146117,Some-college,10,Never-married,Machine-op-inspct,Not-in-family,White,Male,0,0,42,United-States,<=50K                                   
4 16283,24,Private,138938,Some-college,10,Divorced,Adm-clerical,Not-in-family,White,Female,0,0,40,United-States,<=50K                                           
5 16284,57,Self-emp-inc,258883,HS-grad,9,Married-civ-spouse,Transport-moving,Husband,White,Male,5178,0,60,Hungary,>50K

<img src="img/surprised_face.jpg" width="300"/>

In [3]:
adult <- read.csv("dados/adults.csv", header=TRUE)
head(adult)

Id    age workclass    fnlwgt education    education.num marital.status    
1 16280 34  Private      204991 Some-college 10            Divorced          
2 16281 58  Local-gov    310085 10th          6            Married-civ-spouse
3 16282 25  Private      146117 Some-college 10            Never-married     
4 16283 24  Private      138938 Some-college 10            Divorced          
5 16284 57  Self-emp-inc 258883 HS-grad       9            Married-civ-spouse
6 16285 57  Private      163047 HS-grad       9            Married-civ-spouse
  occupation        relationship  race  sex    capital.gain capital.loss
1 Exec-managerial   Own-child     White Male      0         0           
2 Transport-moving  Husband       White Male      0         0           
3 Machine-op-inspct Not-in-family White Male      0         0           
4 Adm-clerical      Not-in-family White Female    0         0           
5 Transport-moving  Husband       White Male   5178         0           
6 Exec-managerial   Wife          White Female    0         0           
  hours.per.week native.country income
1 44             United-States  <=50K 
2 40             United-States  <=50K 
3 42             United-States  <=50K 
4 40             United-States  <=50K 
5 60             Hungary        >50K  
6 38             United-States  <=50K

## <a id = "parte2"></a><span style="color: orange;">  Parte 2 -  Estatística Básica em R </span>

Agora que já vimos o básico de importação de dados em **R**, estamos prontos para entender como essa linguagem é utilizada efetivamente para fazer análises estatísticas!

Mas não se preocupe, vamos explorando os diversos temas de estatística enquanto aprendemos mais sobre **R**.

## <a id="tipos_de_dados"></a> Tipos de Dados

Um dos pontos base da Estatística consiste em entender nossas informações e como elas elas são organizadas: os **dados**. Esses dados podem ser divididos em dois tipos diferentes: dados **qualitativos** e dados **quantitativos**.

### <a id="quali"></a>Dados Qualitativos

Esses dados representam **categorias** do nosso espaço, ou seja, eles dividem os indivíduos em diversas classes que não se intersectam. Por exemplo, a cor de uma flor é um tipo de dado qualitativo, já que as categoriza em classes como: *'amarelo'*, *'rosa'*, *'vermelho'*, *'azul'*, etc. Esses dados são bem úteis para representar características que não são facilmente descritas por números, como qualidades (daí vem o nome).

In [4]:
# Montando um Dataframe de Cores de Flores

Flor = c("Margarida", "Hortênsia", "Violeta", "Rosa")
Cor = c("Amarelo", "Azul", "Azul", "Vermelho")
flores = data.frame(Flor, Cor)
flores

Flor      Cor     
1 Margarida Amarelo 
2 Hortênsia Azul    
3 Violeta   Azul    
4 Rosa      Vermelho

In [5]:
# Mostrando as flores de cor Azul

flores[flores$Cor == "Azul",]

Flor      Cor 
2 Hortênsia Azul
3 Violeta   Azul

### <a id="quant"></a> Dados Quantitativos

São tipos de dados **numéricos**, ou seja, são mensuráveis e podem ser manipulados matematicamente. Essa categoria engloba os dados representáveis por números, como a altura de uma pessoa ou a área de um terreno. Esses dados são mais versáteis, já que conseguimos fazer comparações, médias e outros tipos de operações aritméticas que não são possíveis com os dados qualitativos.

In [6]:
# Montando um Dataframe de Altura de Pessoas

Nome = c("Marcela", "Alexandre", "Rafaela", "Mariana", "Victor")
Altura = c(161, 184, 172, 174, 169)
alturas = data.frame(Nome, Altura)
alturas

Nome      Altura
1 Marcela   161   
2 Alexandre 184   
3 Rafaela   172   
4 Mariana   174   
5 Victor    169

In [7]:
# Selecionando as pessoas com altura maior que 170 cm

alturas[alturas$Altura > 170,]

Nome      Altura
2 Alexandre 184   
3 Rafaela   172   
4 Mariana   174

## <a id="descritiva"></a>Estatística Descritiva

O tipo de análise estatística mais básico é a **Estatística Descritiva**, que busca simplesmente resumir e descrever os nossos dados da melhor maneira possível. Esse ramo utiliza de diversas medidas para expressar características diferentes dos nossos dados, como veremos a seguir.

### <a id="tendencia_central"></a>Medidas de Tendência Central

As **Medidas de Tendência Central** procuram representrar um valor central em torno do qual os dados se agrupam. Você provavelmente já ouviu falar de alguns exemplos desse tipo de medida, como a **média** e a **mediana**.

#### <a id="media"></a> Média

A **Média** é a medida numérica que representa a posição central dos nossos dados. Seu valor é obtido ao dividir a soma de todos os nossos dados pela sua quantidade.

$$\large \bar x ={\frac {1}{N}}\sum _{i=1}^{N}x_{i}$$

In [8]:
# Média de horas trabalhadas dos indivíduos

mean(adult$hours.per.week)

[1] 40.43747

In [9]:
# Média de idade dos indivíduos divorciados e não-casados

media_a = mean(adult$age[adult$marital.status == "Divorced"])
media_b = mean(adult$age[adult$marital.status == "Never-married"])

# As pessoas divorciadas são, na média, mais velhas que as que nunca casaram

cat(media_a, " > ", media_b)

43.04164  >  28.14997

In [10]:
# Não é possível calcular a média de dados qualitativos

mean(adult$marital.status)

Warning message in mean.default(adult$marital.status):
“argument is not numeric or logical: returning NA”


[1] NA

#### <a id="mediana"></a>Mediana

A **Mediana** é a medida numérica que representa o dado que se encontra na metade da nossa distribuição quanto esta se encontra ordenada. Ou seja, a mediana é o valor que divide os nossos dados de tal forma que metade deles sejam maiores que a mediana e a outra metade seja menor.

In [11]:
# Mediana do grau de escolaridade dos trabalhadores do setor privado

median(adult$education.num[adult$workclass == "Private"])

[1] 10

In [12]:
# Não é possível calcular a mediana de dados qualitativos

median(adult$workclass)

ERROR: Error in median.default(adult$workclass): need numeric data


#### <a id="percentil"></a>Percentil

O **Percentil** não é uma medida de tendência central, mas se relaciona bastante com a Mediana. Os percentis são medidas que dividem as nossas amostras tal que **p%** delas sejam menores que o percentil. O **50º percentil**, portanto, é a nossa já conhecida mediana, enquanto o **25º percentil** é denominado primeiro quartil.

In [13]:
# Obtendo os quartis das horas trabalhadas pelos indivíduos

quartis = quantile(adult$hours.per.week)

# Transpondo o resultado para ficar mais visualizável
t(quartis)

0% 25% 50% 75% 100%
[1,] 1  40  40  45  99

In [14]:
# Obtendo os percentis 30%, 60% e 90% das horas trabalhadas pelas mulheres

percentis = quantile(adult$hours.per.week[adult$sex == "Female"], c(.30, .60, .90))

# Transpondo o resultado para ficar mais visualizável
t(percentis)

30% 60% 90%
[1,] 35  40  48

#### <a id="summary"></a> Summary

Uma maneira bem simples de mostrar vários dos valores que vimos até agora é utilizando a função ```summary()```. Ela nos apresenta o *mínimo*, o *1º quartil*, a *mediana*, a *média*, o *3º quartil* e o *máximo* da variável em questão, da seguinte forma:

In [15]:
# Resumo do grau de escolaridade dos indivíduos

summary(adult$education.num)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1.00    9.00   10.00   10.08   12.00   16.00 

### <a id="variacao"></a> Medidas de Variação

As **Medidas de Variação** medem o *grau de afastamento* dos dados em torno de um valor central. Elas são importantes para medir a homogeneidade dos dados, indicando se nossas amostras estão próximas entre si ou muito afastadas.

#### <a id="amplitude"></a> Amplitude

A medida mais simples de variação é a **Amplitude**, que mede simplesmente a diferença entre a maior e a menor amostra. Entretanto, essa medida é bem sensível outliers em suas extremidades.

$$ \large A =  x_{max} - x_{min} $$

In [16]:
# Calculando a amplitude de idade dos homens

amplitude = max(adult$age[adult$sex == "Male"]) - min(adult$age[adult$sex == "Male"])

amplitude

[1] 73

#### <a id="amplitude_interquartil"></a> Amplitude Interquartil

Uma outra medida de variação mais precisa é a **Amplitude Interquartil**, definida como a diferença entre o 3º e o 1º quartil. Por desprezar os valores mais extremos, essa medida é menos sujeita a flutuações que a amplitude normal.

In [17]:
# Calculando a amplitude interquartil da idade dos homens

IQR(adult$age[adult$sex == "Male"])

[1] 19

#### <a id="variancia"></a> Variância

A **Variância** é uma medida de variação que mensura a dispersão das amostras com relação à média.

Seu valor é obtido a partir do cálculo da média do quadrado das diferenças entre os dados e a média, da seguinte forma:
<br>
<br>

$$\large s^2 = {\frac{\sum_{i=1}^{N}(x_{i} -\bar{x})^2}{N-1}}$$

In [18]:
# Calculando a variância das horas trabalhadas de quem possui mais de 30 anos

var(adult$hours.per.week[adult$age > 30])

[1] 143.0836

#### <a id="desv_pad"></a> Desvio Padrão

O **Desvio Padrão** é uma outra medida de variação obtida simplesmente pela raiz quadrada da variância. Essa raiz é feita para que a unidade de medida seja a mesma dos nossos dados.
<br>
<br>

$$\large s = \sqrt{\frac{\sum_{i=1}^{N}(x_{i} -\bar{x})^2}{N-1}}$$

In [19]:
# Calculando o desvio padrão das horas trabalhadas de quem possui mais de 30 anos

sd(adult$hours.per.week[adult$age > 30])

[1] 11.96176

#### <a id="covar"></a> Covariância

A **Covariância** é a medida da *relação linear* entre duas variáveis: seu valor representa o quanto elas "caminham juntas". Se a covarância é positiva, significa que as duas crescem juntas; se a covariância é negativa, significa que conforme uma cresce, a outra decresce.

Seu valor é obtido da seguinte forma:
<br>

$$\large s_{x,y}=\frac{\sum_{i=1}^{N}(x_{i}-\bar{x})(y_{i}-\bar{y})}{N-1}$$

In [20]:
# Calculando a covarância entre o grau de escolaridade e as horas trabalhadas dos indivíduos

cov(adult$education.num, adult$hours.per.week)

[1] 4.705522

A covariância positiva significa que quanto maior a idade do indivíduo, maior é a chance dele trabalhar mais horas

#### <a id="ccor"></a>Coeficiente de Correlação

O **Coeficiente de Correlação** também mede a relação linear entre duas variáveis, mas é obtida a partir da divisão da covariância pelos desvios padrões individuais de cada variável. Dessa forma, encontramos uma medida **normalizada** da relação entre as duas. 
<br>
<br>

$$\large r_{x,y}= \frac{s_{x,y}}{s_{x} s_{y}}$$

Quanto mais próximo de 1, maior a _relação linear positiva_ entre as duas variáveis, e seu gráfico se aproxima de uma linha reta _crescente_. Quanto mais próximo de -1, maior a _relação linear negativa_ entre as duas variáveis, e seu gráfico se aproxima de uma linha reta _decrescente_. Quando seu valor é próximo de 0, não existe uma relação muito grande entre as duas variáveis.

In [21]:
# Calculando o coeficiente de correlação entre o grau de escolaridade e as horas trabalhadas dos indivíduos

cor(adult$education.num, adult$hours.per.week)

[1] 0.1481269

O grau de escolaridade dos indivíduos tem uma relação linear positiva fraca com as horas semanais de trabalho.

## <a id = "parte3"></a><span style="color: orange;"> Parte 3 - Manipulação de dados em R </span>

A linguagem R tem como um de seus grandes destaques o pacote Tidyverse.<br>
Para falar a verdade, o Tidyverse não é um pacote em si, mas sim um conjunto de pacotes.<br>
Dentro dele, encotramos o `dplyr`, que é instaldo automaticamente com o `install.packages("tidyverse")`.
<br>
Aqui uma [Cheat Sheet do dplyr](https://github.com/rstudio/cheatsheets/blob/master/data-transformation.pdf)
<br>

In [22]:
library(dplyr)


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




### <a id="dados_faltantes"></a>Dados faltantes
<br>
Dados faltantes podem ser um grande problema quando estamos lidando com uma base de dados. Assim, uma abordagem geralmente aplicada é simplesmente removê-los da sua base de dados. No R, muitas funções (a grande maioria delas) exibe um parâmetro `na.rm` que abrevia "remover NAs (dados faltantes)".

In [23]:
vetor <- c(2,3,4,5,NA,7) 
#definimos aqui um vetor, vamos ver o que acontece quando aplicamos a função mean()  a ele
print(mean(vetor))

[1] NA


In [24]:
#Agora com o na.rm = TRUE
print(mean(vetor, na.rm=TRUE))

[1] 4.2


Outra coisa que podemos querer fazer é inputar algum dado padrão aos dados faltantes, isso pode ser feito da seguinte maneira:

In [25]:
vetor[is.na(vetor)] <- 0
print(vetor)

[1] 2 3 4 5 0 7


Vamos dar uma olhada na base adult, vamos ver se ela tem algum dado faltante, messa base eles estão representados como "?" então temos que arrumar isso.

In [44]:
adult[adult == "?"] <- NA

In [47]:
sum(is.na(adult))

[1] 4262

Temos cerca de 4262 dados faltantes na base

### <a id="dplyr"></a>Deep dive in dplyr
Vamos agora nos aprofundar na biblioteca `dplyr`. <br>
Essa biblioteca é boa porque deixa o nosso código extremamente legível para um ser humano.<br>
Isso ocorre porque ela usa uma estrutura de "verbos" que regem o que será feito com a base de dados.<br>
Os verbos existentes na biblioteca são os seguintes:
1. `mutate()` - que adiciona novas variáveis (colunas) ao dataset a partir de operações/ aplicação de funções nas variáveis existentes 
2. `select()` - seleciona as colunas com base em critérios de nome
3. `filter()` - filtra os dados com base em valores.
4. `summarise()` - reduz um grande número de valores para valores resumo (como média, contagem, desvio, etc). Para quem mexe com excel, aqui seria algo para criar uma tabela dinâmica (pivot table). Em geral é usado com `group_by()`
5. `arrange()`- muda a ordenação das linhas com base em algum critério
<br>
<br>
Todos esses verbos podem ser performados tendo como critério um grupo, usando o `group_by()` , mas é mais comum usar isso no `summarise()`. <br>

A estrutura de utilização dos verbos é bem simples: <br>
`base de dados %>% verbo` <br>
Podemos aplicar vários verbos em seguida também, com a seguinte estutura (a ordem de execução será a ordem desses verbos): <br>
`base de dados %>% verbo1 %>% verbo2 %>% verbo3`

#### <a id="mutate"></a> `mutate()` - adicionando variáveis

In [28]:
adult_nova_var <- adult %>% mutate(
            old = if_else( age >= 65, TRUE, FALSE, missing = NULL),
            # a função if_else tem quatro argumentos. O primeiro é a 'afirmação lógica' a ser avaliada
            # no caso, essa afirmação é verificar se a idade é maior que 65 anos 
            # o segundo argumento é o que será inputado no caso da afirmação ser verdadeira
            # o terceiro, o que será inputado no caso de falsa
            # e o quarto, o que será inputado no caso de missing value (valor faltante)
            )
adult_nova_var[1:5,]

Id    age workclass    fnlwgt education    education.num marital.status    
1 16280 34  Private      204991 Some-college 10            Divorced          
2 16281 58  Local-gov    310085 10th          6            Married-civ-spouse
3 16282 25  Private      146117 Some-college 10            Never-married     
4 16283 24  Private      138938 Some-college 10            Divorced          
5 16284 57  Self-emp-inc 258883 HS-grad       9            Married-civ-spouse
  occupation        relationship  race  sex    capital.gain capital.loss
1 Exec-managerial   Own-child     White Male      0         0           
2 Transport-moving  Husband       White Male      0         0           
3 Machine-op-inspct Not-in-family White Male      0         0           
4 Adm-clerical      Not-in-family White Female    0         0           
5 Transport-moving  Husband       White Male   5178         0           
  hours.per.week native.country income old  
1 44             United-States  <=50K  FALSE
2 40             United-States  <=50K  FALSE
3 42             United-States  <=50K  FALSE
4 40             United-States  <=50K  FALSE
5 60             Hungary        >50K   FALSE

#### <a id="select"></a>`select()` - selecionando valores com base em nome

Esse comando é bem simples, nós podemos selecionar as colunas com o critério baseado no nome. 
<br>Isso pode incluir "começa com": `starts_with(padrão procurado)`, "termina com": `ends_with(padrão procurado)`, "contém": `contains(padrão procurado)`... <br>
Pra quem manja mais, pode usar expressões regulares com `matches()` ou muitos outros critérios que vc pode dar uma conferida [aqui!](https://www.rdocumentation.org/packages/tidyselect/versions/1.0.0/topics/select_helpers)

Vamos selecionar as variávis que contém a palavra "capital"

In [31]:
selecting <- adult %>% select(contains("capital"))
selecting[1:5,]

capital.gain capital.loss
1    0         0           
2    0         0           
3    0         0           
4    0         0           
5 5178         0

#### <a id="filter"></a> `filter()` - Filtrando nossa base de dados!

Vamos supor que você quer analisar sua base de dados e segmentar essa análise com base em algum critério.. Isso é só um dos cenários em que pode ser muito importante você usar filtros!<br>
Vamos aqui filtrar a base para pessoas com educação formal acima da média

In [32]:
filtro_education = adult %>% filter(education.num > mean(education.num, na.rm = TRUE))
filtro_education[1:5,]

Id    age workclass fnlwgt education  education.num marital.status    
1 16287 25  Private   288519 Bachelors  13            Never-married     
2 16291 42  Private   213214 Assoc-acdm 12            Married-civ-spouse
3 16292 24  State-gov 215797 Bachelors  13            Never-married     
4 16295 24  Local-gov 103277 Bachelors  13            Never-married     
5 16299 22  Private   230704 Assoc-acdm 12            Never-married     
  occupation        relationship  race  sex    capital.gain capital.loss
1 Adm-clerical      Not-in-family White Female 0            0           
2 Sales             Husband       White Male   0            0           
3 Tech-support      Not-in-family White Female 0            0           
4 Prof-specialty    Own-child     White Female 0            0           
5 Handlers-cleaners Own-child     Black Male   0            0           
  hours.per.week native.country income
1 40             United-States  <=50K 
2 40             United-States  <=50K 
3 40             United-States  <=50K 
4 50             United-States  <=50K 
5 20             Jamaica        <=50K

#### <a id="summarise"></a> `summarise()` - "Tabelas dinâmicas" no R

Usar o verbo `summarise()` vai produzir um único número com a(s) medida(s) resumo escolhidas. <br>
Mas como, muitas vezes, nós queremos segmentar a análise para subgrupos diferentes do dataset, o `summarise()` costuma ser utilizado com o `group_by()`. <br>
O `group_by()` vai fazer com que as medidas resumos sejam feitas com base nos diferentes valores tidos em uma coluna.

Vamos, primeiro, usar o `summarise()` puro. Perceba que ele devolve a medida escolhida levando em consideração todas as linhas.

In [36]:
adult %>%
  summarise(media_idade = mean(age), media_anos_educacao = mean(education.num))

media_idade media_anos_educacao
1 38.58163    10.08059

Agora, utilizando o `group_by()` podemos segmentar os números com base em uma ou mais categorias.

In [38]:
adult %>% 
    group_by(race, sex) %>%
    summarise(media_idade = mean(age), media_anos_educacao = mean(education.num))

race               sex    media_idade media_anos_educacao
1  Amer-Indian-Eskimo Female 37.11765     9.697479          
2  Amer-Indian-Eskimo Male   37.20833     9.072917          
3  Asian-Pac-Islander Female 35.08960    10.390173          
4  Asian-Pac-Islander Male   39.07359    11.245310          
5  Black              Female 37.85402     9.549839          
6  Black              Male   37.68260     9.423199          
7  Other              Female 31.67890     8.899083          
8  Other              Male   34.65432     8.802469          
9  White              Female 36.81162    10.127980          
10 White              Male   39.65253    10.138372

Perceba que nos é devolvida uma medida para cada combinação possível de grupos

#### <a id="arrange"></a> `arrange()` - Alterando a ordem das linhas do seu dataset

Esse é provavelmente o verbo mais fácil de ser usado. Você passa como argumento um ou mais critérios de ordenação (sendo que a ordem de preferência é a ordem passada). <br>
Caso queira que a ordenação seja decrescente, é só utilizar a função `desc()`.

In [40]:
arrange_adult = adult %>% arrange(desc(age), education.num)
arrange_adult[1:5,]

Id    age workclass fnlwgt education education.num marital.status    
1 46769 90  ?         166343 1st-4th   2             Widowed           
2 26946 90  Private   226968 7th-8th   4             Married-civ-spouse
3 34834 90  ?         175444 7th-8th   4             Separated         
4 48811 90  Local-gov 214594 7th-8th   4             Married-civ-spouse
5 25836 90  Private   141758 9th       5             Never-married     
  occupation        relationship  race  sex    capital.gain capital.loss
1 ?                 Not-in-family Black Female    0         0           
2 Machine-op-inspct Husband       White Male      0         0           
3 ?                 Not-in-family White Female    0         0           
4 Protective-serv   Husband       White Male   2653         0           
5 Adm-clerical      Not-in-family White Female    0         0           
  hours.per.week native.country income
1 40             United-States  <=50K 
2 40             United-States  <=50K 
3 15             United-States  <=50K 
4 40             United-States  <=50K 
5 40             United-States  <=50K

### <a id="varios_verbos"></a> Usando vários verbos ao mesmo tempo

Uma coisa legal do `dplyr` é que a gente pode usar vários verbos ao mesmo tempo

Aqui vamos pegar uma base apenas com aqueles com anos de educação acima da média, depois vamos avaliar as combinações de raça e sexo possíves, para cada uma dando sua média de idade e média de anos de educação. Finalmente, vamos fazer com que os resultados sejam ordenados da maior média de anos de educação para a menor

In [49]:
adult %>% 
    filter(education.num > mean(education.num, na.rm = TRUE)) %>%
    group_by(race, sex) %>%
    summarise(media_idade = mean(age), media_anos_educacao = mean(education.num)) %>%
    arrange(desc(media_anos_educacao))

race               sex    media_idade media_anos_educacao
1  Asian-Pac-Islander Male   39.71467    13.41033           
2  Other              Male   36.75000    13.16667           
3  White              Male   41.53804    13.09590           
4  Asian-Pac-Islander Female 35.61379    12.98621           
5  White              Female 37.86561    12.89654           
6  Black              Male   39.99690    12.85139           
7  Other              Female 35.70833    12.79167           
8  Amer-Indian-Eskimo Male   39.62963    12.74074           
9  Black              Female 37.30769    12.60355           
10 Amer-Indian-Eskimo Female 37.03226    12.32258

## <a id = "parte3"></a><span style="color: orange;"> Parte 4 - Visualização de dados em R </span>

## <a id = "parte5"></a><span style="color: orange;"> Parte 5 - Inferência Estatística em R </span>